File: S1/Q2.py
import hashlib
import time
import itertools
from concurrent.futures import ThreadPoolExecutor, as_completed
import statistics

def create_table(words):
    table = {}
    def process_word(item):
        stripped_item = item.strip()
        local_permutations = case_permutations(stripped_item)
        local_table = {}
        for perm in local_permutations:
            hashed_item = hashlib.sha1(perm.encode()).hexdigest()
            local_table[perm] = hashed_item
        return local_table

    with ThreadPoolExecutor() as executor:
        futures = [executor.submit(process_word, item) for item in words]
        for future in as_completed(futures):
            local_table = future.result()
            table.update(local_table)
    return table

def case_permutations(word):
    substitutions = {
        'a': '@', 'i': '1!', 'l': '1Ii', 'o': '0', 's': '$5'
    }
    
    def char_variations(char):
        lower = char.lower()
        if lower in substitutions:
            return char.lower() + char.upper() + substitutions[lower]
        return char.lower() + char.upper()

    variations = map(char_variations, word)
    return (''.join(combo) for combo in itertools.product(*variations))

def run_trial(words):
    start_time = time.time()
    rainbow_table = create_table(words)
    end_time = time.time()
    creation_time = end_time - start_time
    table_size = len(rainbow_table)
    table_memory_size = sum(
        len(key.encode()) + len(value.encode()) for key, value in rainbow_table.items()
    )
    return creation_time, table_size, table_memory_size

def main():
    filename = "10k-most-common.txt"
    words = []
    try:
        with open(filename, "r") as file:
            words = file.readlines()
    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
        exit(1)

    num_trials = 10
    creation_times = []
    table_sizes = []
    table_memory_sizes = []

    for i in range(num_trials):
        print(f"Running trial {i+1}/{num_trials}")
        creation_time, table_size, table_memory_size = run_trial(words)
        creation_times.append(creation_time)
        table_sizes.append(table_size)
        table_memory_sizes.append(table_memory_size)

    avg_creation_time = statistics.mean(creation_times)
    avg_table_size = statistics.mean(table_sizes)
    avg_table_memory_size = statistics.mean(table_memory_sizes)

    print(f"\nAverage Results over {num_trials} trials:")
    print(f"Average rainbow table creation time: {avg_creation_time:.4f} seconds")
    print(f"Average table size: {avg_table_size:.0f} entries")
    print(f"Average table memory size: {avg_table_memory_size / (1024 * 1024):.2f} MB")

if __name__ == "__main__":
    main()

File: S1/Q2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <ctype.h> 
#include <openssl/sha.h>

#define MAX_WORD_LENGTH 100
#define MAX_PERMUTATIONS 100000000
#define MAX_WORDS 100000000

typedef struct {
    char *key;
    char *value;
} KeyValuePair;

typedef struct {
    KeyValuePair *data;
    size_t size;
    size_t capacity;
} HashMap;

void init_hash_map(HashMap *map, size_t initial_capacity) {
    map->data = malloc(initial_capacity * sizeof(KeyValuePair));
    map->size = 0;
    map->capacity = initial_capacity;
}

void insert_hash_map(HashMap *map, const char *key, const char *value) {
    if (map->size >= map->capacity) {
        map->capacity *= 2;
        map->data = realloc(map->data, map->capacity * sizeof(KeyValuePair));
    }
    map->data[map->size].key = strdup(key);
    map->data[map->size].value = strdup(value);
    map->size++;
}

void free_hash_map(HashMap *map) {
    for (size_t i = 0; i < map->size; i++) {
        free(map->data[i].key);
        free(map->data[i].value);
    }
    free(map->data);
}

void case_permutations(const char *word, char **permutations, size_t *perm_count) {
    static const char *substitutions[] = {"@", "1!", "1Ii", "0", "$5"};
    static const char substitution_chars[] = {'a', 'i', 'l', 'o', 's'};
    
    size_t word_len = strlen(word);
    char variations[MAX_WORD_LENGTH][5];
    size_t variation_counts[MAX_WORD_LENGTH];

    for (size_t i = 0; i < word_len; i++) {
        char c = word[i];
        char lower = tolower(c);
        int found = 0;
        for (int j = 0; j < 5; j++) {
            if (lower == substitution_chars[j]) {
                snprintf(variations[i], 5, "%c%c%s", tolower(c), toupper(c), substitutions[j]);
                variation_counts[i] = strlen(variations[i]);
                found = 1;
                break;
            }
        }
        if (!found) {
            snprintf(variations[i], 3, "%c%c", tolower(c), toupper(c));
            variation_counts[i] = 2;
        }
    }

    *perm_count = 1;
    for (size_t i = 0; i < word_len; i++) {
        *perm_count *= variation_counts[i];
    }

    for (size_t i = 0; i < *perm_count; i++) {
        size_t temp = i;
        for (size_t j = 0; j < word_len; j++) {
            permutations[i][j] = variations[j][temp % variation_counts[j]];
            temp /= variation_counts[j];
        }
        permutations[i][word_len] = '\0';
    }
}

void create_table(char **words, size_t word_count, HashMap *table) {
    char **permutations = malloc(MAX_PERMUTATIONS * sizeof(char *));
    for (size_t i = 0; i < MAX_PERMUTATIONS; i++) {
        permutations[i] = malloc(MAX_WORD_LENGTH * sizeof(char));
    }

    for (size_t i = 0; i < word_count; i++) {
        size_t perm_count;
        case_permutations(words[i], permutations, &perm_count);

        for (size_t j = 0; j < perm_count; j++) {
            unsigned char hash[SHA_DIGEST_LENGTH];
            char hash_hex[SHA_DIGEST_LENGTH * 2 + 1];
            SHA1((unsigned char *)permutations[j], strlen(permutations[j]), hash);
            
            for (int k = 0; k < SHA_DIGEST_LENGTH; k++) {
                sprintf(&hash_hex[k * 2], "%02x", hash[k]);
            }
            
            insert_hash_map(table, permutations[j], hash_hex);
        }
    }

    for (size_t i = 0; i < MAX_PERMUTATIONS; i++) {
        free(permutations[i]);
    }
    free(permutations);
}

int main() {
    FILE *file = fopen("10k-most-common.txt", "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    char **words = malloc(MAX_WORDS * sizeof(char *));
    for (int i = 0; i < MAX_WORDS; i++) {
        words[i] = malloc(MAX_WORD_LENGTH * sizeof(char));
    }

    size_t word_count = 0;
    while (fgets(words[word_count], MAX_WORD_LENGTH, file) && word_count < MAX_WORDS) {
        words[word_count][strcspn(words[word_count], "\n")] = 0;
        word_count++;
    }
    fclose(file);

    int num_trials = 10;
    double total_time = 0;
    size_t total_size = 0;
    size_t total_memory = 0;

    for (int i = 0; i < num_trials; i++) {
        printf("Running trial %d/%d\n", i + 1, num_trials);

        HashMap table;
        init_hash_map(&table, 100000000);

        clock_t start = clock();
        create_table(words, word_count, &table);
        clock_t end = clock();

        double creation_time = ((double) (end - start)) / CLOCKS_PER_SEC;
        size_t table_size = table.size;
        size_t table_memory = 0;
        for (size_t j = 0; j < table.size; j++) {
            table_memory += strlen(table.data[j].key) + strlen(table.data[j].value) + 2;
        }

        total_time += creation_time;
        total_size += table_size;
        total_memory += table_memory;

        free_hash_map(&table);
    }

    double avg_creation_time = total_time / num_trials;
    double avg_table_size = (double)total_size / num_trials;
    double avg_table_memory = (double)total_memory / num_trials;

    printf("\nAverage Results over %d trials:\n", num_trials);
    printf("Average rainbow table creation time: %.4f seconds\n", avg_creation_time);
    printf("Average table size: %.0f entries\n", avg_table_size);
    printf("Average table memory size: %.2f MB\n", avg_table_memory / (1024.0 * 1024.0));
    printf("Average Performance: %f seconds per entry\n", avg_creation_time / avg_table_size);

    for (int i = 0; i < MAX_WORDS; i++) {
        free(words[i]);
    }
    free(words);

    return 0;
}